File: csharp/CollatzConjecture.cs
using System;

public static class CollatzConjecture
{
    public static int Steps(int number)
    {
        if (number <= 0)
        {
            throw new ArgumentException();
        }

        return Collatz(number, 0);
    }

    public static int NextCollatzStep(int n) => n % 2 == 0 ? (n / 2) : (3 * n + 1);

    public static int Collatz(int m, int count) => m == 1 ? count : Collatz(NextCollatzStep(m), count + 1);
}

File: c/collatz_conjecture.c
#include "collatz_conjecture.h"

int steps(int start) {
  if (start <= 0) {
    return -1;
  }
  int count_steps = 0;
  while (start != 1) {
    if (start % 2 == 0) {
      start = start / 2;
    } else {
      start = 3 * start + 1;
    }
    count_steps++;
  }

  return count_steps;
}
